File: Divide and Conquer/boj_2630.py
import sys

input = sys.stdin.readline


def recursion(x, y, m):
    global white, blue
    for i in range(x, x + m):
        for j in range(y, y + m):
            if board[i][j] != board[x][y]:
                recursion(x, y, m // 2)
                recursion(x, y + m // 2, m // 2)
                recursion(x + m // 2, y, m // 2)
                recursion(x + m // 2, y + m // 2, m // 2)
                return
    if board[x][y]:
        blue += 1
    else:
        white += 1


n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

white = blue = 0
recursion(0, 0, n)

print(white)
print(blue)

File: Divide and Conquer/boj_2630.cpp
#include <iostream>
using namespace std;

int white = 0, blue = 0;
int board[128][128];

void recursion(int x, int y, int n) {
	for (int i = x; i < x + n; i++) {
		for (int j = y; j < y + n; j++) {
			if (board[i][j] != board[x][y]) {
				recursion(x, y, n/2);
				recursion(x + n/2, y, n/2);
				recursion(x, y + n/2, n/2);
				recursion(x + n/2, y + n/2, n/2);
				return;
			}
		}
	}
	
	if (board[x][y])
		blue++;
	else
		white++;
}

int main() {
	int n;

	cin >> n;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> board[i][j];
		}
	}

	recursion(0, 0, n);

	cout << white << "\n";
	cout << blue << "\n";

	return 0;
}
